### PR TITLE
742 builtin story rules

### PR DESCRIPTION
**Pull request by francoisno**
@vsct-jburet : L'implémentation côté bot-engine/bot-api me semble un peu bricolo et j'ai un peu peur d'un impact sur les bots en mode Intégré (j'ai fait mes tests en Bot Api).
Si ça te paraît correct, j'ajouterai quelques TU pour compléter.

### CODE REVIEW

File: bot/engine/src/main/kotlin/engine/config/BotDefinitionWrapper.kt [hunk 1]
```diff
@@ -37,11 +38,20 @@ internal class BotDefinitionWrapper(val botDefinition: BotDefinition) : BotDefin
     @Volatile
     private var allStories: List<StoryDefinition> = botDefinition.stories
 
-    fun updateStories(configuredStories: List<ConfiguredStoryDefinition>) {
-        logger.debug { "refresh configured stories for ${botDefinition.botId}" }
+    private val dao: StoryDefinitionConfigurationDAO = injector.provide()
+
+    @Volatile
+    private var storyConfRedirections: Map<String, String?> = emptyMap()
```
Comment (vsct-jburet) [inline bot/engine/src/main/kotlin/engine/config/BotDefinitionWrapper.kt:44]: storyConfRedirections should be renamed storyRedirectionsMap with a comment to explain the key and value types

File: bot/engine/src/main/kotlin/engine/config/BotDefinitionWrapper.kt [hunk 2]
```diff
@@ -37,11 +38,20 @@ internal class BotDefinitionWrapper(val botDefinition: BotDefinition) : BotDefin
...
+
+    fun updateStories(configuredStories: List<ConfiguredStoryDefinition>, applicationId: String?) {
         this.configuredStories = configuredStories.filter { it.answerType != builtin }.groupBy { it.id }
         //configured stories can override built-in
         allStories = (this.configuredStories + botDefinition.stories.groupBy { it.id }).values.flatten()
+        val storyConfigurations = with(botDefinition) {
```
Comment (vsct-jburet) [inline bot/engine/src/main/kotlin/engine/config/BotDefinitionWrapper.kt:50]: Why do you retrieve storyConfigurations from databases and do not use configuredStories param?

File: bot/engine/src/test/kotlin/engine/config/ConfiguredStoryHandlerTest.kt [hunk 3]
```diff
@@ -51,34 +51,6 @@ class ConfiguredStoryHandlerTest {
         override val connectorType: ConnectorType = ConnectorType.none
     ) : ConnectorMessage
 
-    @Test
```
Comment (vsct-jburet) [inline bot/engine/src/test/kotlin/engine/config/ConfiguredStoryHandlerTest.kt:54]: please reuse this test :)

File: bot/admin/web/src/app/bot/feature/story-rule.component.ts [hunk 4]
```diff
@@ -32,7 +32,9 @@ export class StoryRuleComponent implements OnInit {
   create: boolean = false;
   loadingStoryRules: boolean = false;
   stories: StoryDefinitionConfiguration[] = [];
-  features: StoryFeature[] = [];
+//   features: StoryFeature[] = [];
```
Comment (vsct-jburet) [inline bot/admin/web/src/app/bot/feature/story-rule.component.ts:35]: remove

File: bot/engine/src/main/kotlin/engine/config/BotDefinitionWrapper.kt [hunk 5]
```diff
@@ -37,11 +38,20 @@ internal class BotDefinitionWrapper(val botDefinition: BotDefinition) : BotDefin
...
+            dao.getStoryDefinitionsByNamespaceAndBotId(namespace, botId).map { story -> story.storyId to story }.toMap()
+        }
+        storyConfRedirections = storyConfigurations.map { it.key to followStoryRedirects(applicationId, storyConfigurations, it.key, null) }.toMap()
+        logger.info("Story rules for '${botDefinition.botId} ($applicationId)': $storyConfRedirections")
```
Comment (vsct-jburet) [inline bot/engine/src/main/kotlin/engine/config/BotDefinitionWrapper.kt:54]: please use debug log level for this and { }

File: bot/engine/src/main/kotlin/engine/config/BotDefinitionWrapper.kt [hunk 6]
```diff
@@ -59,24 +69,41 @@ internal class BotDefinitionWrapper(val botDefinition: BotDefinition) : BotDefin
...
+                val theStory = stories[this]
+                if (theStory?.isDisabled(applicationId) == true) { return null }
+                theStory?.findFeatures(applicationId)?.find { it.enabled && !it.switchToStoryId.isNullOrBlank() }?.let {
+                    followStoryRedirects(applicationId, stories.filterKeys { it -> this != it }, requestStoryId, it.switchToStoryId)
+                } ?: this
+            }
+
+    override fun findStoryDefinition(intent: String?, applicationId: String?): StoryDefinition {
+        val redirectStoryId = storyConfRedirections[intent]
+        if (redirectStoryId != null && redirectStoryId != intent) {
```
Comment (vsct-jburet) [inline bot/engine/src/main/kotlin/engine/config/BotDefinitionWrapper.kt:86]: you can use only one return here

File: bot/engine/src/main/kotlin/engine/config/StoryConfigurationMonitor.kt [hunk 7]
```diff
@@ -54,10 +54,12 @@ internal object StoryConfigurationMonitor {
     }
 
     private fun refresh(bot: Bot) {
+        logger.debug("Refreshing bot ${bot.botDefinition.botId} (${bot.configuration.name})...")
```
Comment (vsct-jburet) [inline bot/engine/src/main/kotlin/engine/config/StoryConfigurationMonitor.kt:57]: use {}

logger.debug{"Refreshing bot ${bot.botDefinition.botId} (${bot.configuration.name})..."}

File: bot/engine/src/main/kotlin/engine/config/BotDefinitionWrapper.kt [hunk 8]
```diff
@@ -37,11 +38,20 @@ internal class BotDefinitionWrapper(val botDefinition: BotDefinition) : BotDefin
...
     private var allStories: List<StoryDefinition> = botDefinition.stories
 
-    fun updateStories(configuredStories: List<ConfiguredStoryDefinition>) {
-        logger.debug { "refresh configured stories for ${botDefinition.botId}" }
+    private val dao: StoryDefinitionConfigurationDAO = injector.provide()
+
+    @Volatile
+    private var storyConfRedirections: Map<String, String?> = emptyMap()
+
+    fun updateStories(configuredStories: List<ConfiguredStoryDefinition>, applicationId: String?) {
```
Comment (vsct-jburet) [inline bot/engine/src/main/kotlin/engine/config/BotDefinitionWrapper.kt:46]: applicationId should be not null

File: bot/engine/src/main/kotlin/engine/config/BotDefinitionWrapper.kt [hunk 9]
```diff
@@ -59,24 +69,41 @@ internal class BotDefinitionWrapper(val botDefinition: BotDefinition) : BotDefin
         return findStoryDefinition(intent?.wrappedIntent()?.name)
     }
 
-    override fun findStoryDefinition(intent: String?, applicationId: String?): StoryDefinition =
-        intent?.let { i ->
+    fun followStoryRedirects(applicationId: String?, stories: Map<String, StoryDefinitionConfiguration>, requestStoryId: String, switchStoryId: String?): String? =
```
Comment (vsct-jburet) [inline bot/engine/src/main/kotlin/engine/config/BotDefinitionWrapper.kt:72]: should be private

File: bot/engine/src/main/kotlin/engine/config/BotDefinitionWrapper.kt [hunk 10]
```diff
@@ -59,24 +69,41 @@ internal class BotDefinitionWrapper(val botDefinition: BotDefinition) : BotDefin
...
+        if (switchStoryId == requestStoryId)
+            requestStoryId
+        else
+            with(switchStoryId ?: requestStoryId) {
+                val theStory = stories[this]
+                if (theStory?.isDisabled(applicationId) == true) { return null }
+                theStory?.findFeatures(applicationId)?.find { it.enabled && !it.switchToStoryId.isNullOrBlank() }?.let {
+                    followStoryRedirects(applicationId, stories.filterKeys { it -> this != it }, requestStoryId, it.switchToStoryId)
```
Comment (vsct-jburet) [inline bot/engine/src/main/kotlin/engine/config/BotDefinitionWrapper.kt:80]: it shadowed

File: bot/engine/src/test/kotlin/engine/config/ConfiguredStoryHandlerTest.kt (hunk 3 above)
Comment (francoisno) [reply bot/engine/src/test/kotlin/engine/config/ConfiguredStoryHandlerTest.kt:54]: Story-rules redirection is now tested by `BotDefinitionWrapperTest`.
(`BotBusTest` still tests the logic with `SWITCH_STORY_BUS_KEY`.)

**Comment by francoisno** [comment]
Review fixes + more bugfixes and cleanup added to the branch.
Ready to merge if your are @vsct-jburet

**Pull request merged**